### PR TITLE
Breaking: convert SourceCode to ES6 class (refs #8231)

### PR DESCRIPTION
**Pull request by not-an-aardvark**
**What is the purpose of this pull request? (put an "X" next to item)**

[x] Other, please explain:

**What changes did you make? (Give an overview)**

This converts SourceCode to an ES6 class. This will not break clients that were using SourceCode as documented, but it could break clients that are relying on undocumented behavior in SourceCode (e.g. enumerable methods).

Also see: https://github.com/eslint/eslint/issues/8231

**Is there anything you'd like reviewers to focus on?**

Nothing in particular

### CODE REVIEW

File: lib/util/source-code.js [hunk 1]
```diff
@@ -106,86 +106,85 @@ function sortedMerge(tokens, comments) {
...
-    return text.split(astUtils.createGlobalLinebreakMatcher());
-};
+        // create token store methods
+        const tokenStore = new TokenStore(ast.tokens, ast.comments);
 
-SourceCode.prototype = {
-    constructor: SourceCode,
+        for (const methodName of TokenStore.PUBLIC_METHODS) {
+            this[methodName] = tokenStore[methodName].bind(tokenStore);
+        }
```
Comment (mysticatea) [inline lib/util/source-code.js]: If this `SourceCode` class extends `TokenStore` class, probably we can remove this copy of TokenStore's methods.

Comment (mysticatea) [inline lib/util/source-code.js]: (in that case, TokenStore's private members should be initialized with `Symbol`s, though.)

Comment (not-an-aardvark) [inline lib/util/source-code.js]: It looks like `TokenStore` doesn't actually have any private methods right now.

```js
Object.getOwnPropertyNames(TokenStore.prototype)
  .filter(name => !TokenStore.PUBLIC_METHODS.includes(name))

// => [ 'constructor' ]
```

Comment (mysticatea) [inline lib/util/source-code.js]: Yes. But `TokenStore` has 3 private fields, `tokens`, `comments`, and `indexMap`.

Comment (not-an-aardvark) [inline lib/util/source-code.js]: Good point. Ok, I'll use symbols for the private fields.

**Comment by ilyavolodin** [comment]
@gyandeeps Do you still require changes or could this be merged?

**Comment by not-an-aardvark** [comment]
Oops, I missed @gyandeeps's review. I'll look into cleaning the comments.

**Pull request merged**